// react-sentry/lib/index.js
"use strict";

module.exports = {
  componentWillMount: function componentWillMount() {
    this.__rs_listeners = [];
  },

  componentWillUnmount: function componentWillUnmount() {
    this.__rs_listeners.forEach(function (listener) {
      var emitter = listener.emitter;
      var eventName = listener.eventName;
      var callback = listener.callback;

      var removeListener = emitter.removeListener || emitter.removeEventListener;
      removeListener.call(emitter, eventName, callback);
    });
  },

  watch: function watch(emitter, eventName, callback) {
    this.__rs_listeners.push({
      emitter: emitter,
      eventName: eventName,
      callback: callback
    });

    var addListener = emitter.addListener || emitter.addEventListener;
    addListener.call(emitter, eventName, callback);
  },

  unwatch: function unwatch(emitter, eventName, callback) {
    this.__rs_listeners = this.__rs_listeners.filter(function (listener) {
      return listener.emitter === emitter && listener.eventName === eventName && listener.callback === callback;
    });

    var removeListener = emitter.removeListener || emitter.removeEventListener;
    removeListener.call(emitter, eventName, callback);
  }
};
